### PR TITLE
Adding a configurable property to JSQMessagesViewController controlli…

### PR DESCRIPTION
**Pull request by fishstix-medhelp**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have squashed my commits into 1 commit.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/.github/CONTRIBUTING.md). Confirmation: __:muscle::sunglasses::facepunch:__

## What's in this pull request?

Adding a configurable property to JSQMessagesViewController controlling whether the typing indicator gets the new animated behavior.  Default is NO, adhearing to previous behavior.
Updating Unit Tests to assert default "NO" values

### CODE REVIEW

File: .github/ISSUE_TEMPLATE.md [hunk 1]
```diff
@@ -41,8 +41,6 @@
 
 > ...
 
-</details>
-
```
Comment (jessesquires) [inline .github/ISSUE_TEMPLATE.md]: let's revert this unrelated change

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.h [hunk 2]
```diff
@@ -61,7 +61,11 @@ FOUNDATION_EXPORT const CGFloat kJSQMessagesTypingIndicatorFooterViewHeight;
  */
 - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
                 messageBubbleColor:(UIColor *)messageBubbleColor
+                          animated:(BOOL)animated
                shouldDisplayOnLeft:(BOOL)shouldDisplayOnLeft
                  forCollectionView:(UICollectionView *)collectionView;
 
 @end
+
+@interface JSQMessagesTypingIndicatorCircleView : UIView
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.h]: let's put this class in a new file

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 3]
```diff
@@ -21,17 +21,31 @@
...
 const CGFloat kJSQMessagesTypingIndicatorFooterViewHeight = 46.0f;
 
+@interface JSQMessagesTypingIndicatorAnimator : NSObject
+- (id) initWithTypingIndicatorView:(JSQMessagesTypingIndicatorFooterView*) view :(UIColor*) ellipsisColor;
+@property (nonatomic, strong) JSQMessagesTypingIndicatorFooterView *typingView;
+@property (nonatomic, strong) UIColor *ellipsisColor;
+@property (nonatomic, assign) BOOL animating;
+- (void) startAnimating;
+- (void) stopAnimating;
+@end
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: once in a new file, let's clean up this formatting to be consistent with the rest of the project :smile:

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 4]
```diff
@@ -109,7 +124,123 @@ - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
...
+}
+
+- (void) setTypingIndicatorEllipsisColor:(UIColor*) ellipsisColor
+{
+    self.typingIndicatorView1.backgroundColor = ellipsisColor;
+    self.typingIndicatorView2.backgroundColor = ellipsisColor;
+    self.typingIndicatorView3.backgroundColor = ellipsisColor;
+}
+
+- (void) setTypingIndicatorAnimated:(BOOL) animated :(UIColor*) ellipsisColor
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: um... there's no parameter name for `ellipsisColor` ??

formatting style:

```objective-c
- (void)setTypingIndicatorAnimated:(BOOL)animated ellipsisColor:(UIColor *)ellipsisColor
```

let's run through all of these additions/changes and make sure style is consistent with the rest of the project.

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 5]
```diff
@@ -109,7 +124,123 @@ - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
...
+{
+    // Configure Base Color
+    [self setTypingIndicatorEllipsisColor:ellipsisColor];
+    // Remove existing animator if exist
+    if (self.animator) {
+        [self.animator stopAnimating];
+    }
+    // If animated -> New Animator
+    if (animated) {
+        self.animator = [[JSQMessagesTypingIndicatorAnimator alloc] initWithTypingIndicatorView:self :ellipsisColor];
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: yeah. this is weird.

should be `initWithTypingIndicatorView:self color:ellipsisColor`

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 6]
```diff
@@ -109,7 +124,123 @@ - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
...
+        [self.animator startAnimating];
+    }
+}
+
+@end
+
+@implementation JSQMessagesTypingIndicatorAnimator
+
+- (id) initWithTypingIndicatorView:(JSQMessagesTypingIndicatorFooterView *)view :(UIColor*) ellipsisColor
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: prefer `instancetype` over `id`

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 7]
```diff
@@ -109,7 +124,123 @@ - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
...
+{
+    if (self = [self init]) {
+        self.typingView = view;
+        self.ellipsisColor = ellipsisColor;
+        self.animating = NO;
+        
+        // Initialize Colors
+        self.typingView.typingIndicatorView1.backgroundColor = ellipsisColor;
+        self.typingView.typingIndicatorView2.backgroundColor = ellipsisColor;
+        self.typingView.typingIndicatorView3.backgroundColor = ellipsisColor;
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: instead of `self.property` use `_property` in `init`. 

example: `_typingView = view;`

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 8]
```diff
@@ -109,7 +124,123 @@ - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
...
+{
+    self.animating = NO;
+}
+
+// How long does each individual dot stay dark for?
+#define kTypingIndicatorSingleDot_Darker_AnimationDuration 1.5f
+// How long does it take to transition from one dark circle to another?
+#define kTypingIndicatorSingleDot_TransitionTime 0.4f
+// Darkening co-efficient
+#define kTypingIndicatorAnimatedDot_Darkening_Coefficient 3.0f
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: no `#define`s

let's either: (1) define constants inline where these are used, or (2) change these to `static const CGFloat`

File: JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m [hunk 9]
```diff
@@ -109,7 +124,123 @@ - (void)configureWithEllipsisColor:(UIColor *)ellipsisColor
...
+
+
+@implementation JSQMessagesTypingIndicatorCircleView
+
+-(id) init {
+    if (self = [super init]) {
+        // Configure Circle
+    }
+    return self;
+}
```
Comment (jessesquires) [inline JSQMessagesViewController/Views/JSQMessagesTypingIndicatorFooterView.m]: can remove this `init` override

**Comment by jessesquires** [comment]
Thanks so much for working on this @fishstix-medhelp ! :+1: 

I've left a few comments to address. In addition, we'll need to do the following:

1. Resolve the merge conflicts
2. Update the PR to merge into `release_8.0` branch instead of `develop`

**Comment by jessesquires** [comment]
Thanks @fishstix-medhelp -- looks some merging issues. Mind squashing into 1 commit?

**Comment by jessesquires** [comment]
Friendly ping @fishstix-medhelp :smile:

**Comment by karthikgollapudi** [comment]
This animated typing indicator looks awesome. Does anyone know when it will get merged in?

**Comment by fishstix-medhelp** [comment]
Hi, I am honestly having a difficult type reconciling the merging issues with the release branch; either collapsing too many commits or not enough, etc...  And unfortunately for anyone else, I'm using this project against my current fork with this feature enabled.  I'm sorry to say this no longer a priority for me.

**Comment by jessesquires** [comment]
Thanks @fishstix-medhelp ! No worries 😄
